File: java/gildedrose_covered/src/test/java/gildedrose/GildedRoseTest.java
package gildedrose;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class GildedRoseTest {

    public static final String BACKSTAGE_PASS = "Backstage passes to a TAFKAL80ETC concert";

    private Item[] createItemArray(String itemName, int sellIn, int quality) {
        return new Item[] { new Item(itemName, sellIn, quality) };
    }

    @Test
    public void LegendaryItem_QualityDoesNotDecrease() {
        // Arrange
        GildedRose app = new GildedRose(createItemArray("Sulfuras, Hand of Ragnaros", 20, 80));

        // Act
        app.updateQuality();

        // Assert
        assertEquals(80, app.items[0].quality, "Quality is not decreased for this legendary item");
    }

    @Test
    public void LegendaryItem_NeverHasToBeSold() {
        GildedRose app = new GildedRose(createItemArray("Sulfuras, Hand of Ragnaros", 1, 80));
        app.updateQuality();
        assertEquals(1, app.items[0].sellIn, "Sellin is not decreased for this legendary item");
    }

    @ParameterizedTest(name = "{0} item SellIn decreases each update")
    @CsvSource({"generic item", "Aged Brie", BACKSTAGE_PASS})
    public void NonLegendaryItem_SellInDate_Decreases(String itemName) {
        GildedRose app = new GildedRose(createItemArray(itemName, 8, 10));
        app.updateQuality();
        assertEquals(7, app.items[0].sellIn, "Item sellin date should decrease by 1 each day");
    }

    @ParameterizedTest(name = "{0} item SellIn will be negative after sellin date reached")
    @CsvSource({"generic item", "Aged Brie", BACKSTAGE_PASS})
    public void NonLegendaryItem_SellinDate_CanBeNegative(String itemName) {
        GildedRose app = new GildedRose(createItemArray(itemName, 0, 25));
        app.updateQuality();
        assertEquals(-1, app.items[0].sellIn, "Sellin date will go negative once sellin date is reached");
    }

    @Test
    public void GenericItem_QualityDecreasesBeforeSellinDate() {
        GildedRose app = new GildedRose(createItemArray("generic item", 5, 10));
        app.updateQuality();
        assertEquals(9, app.items[0].quality, "Item quality should only decrease by 1 each day");
    }

    @Test
    public void GenericItem_QualityDecreasesTwiceAsFastAfterSellinDate() {
        GildedRose app = new GildedRose(createItemArray("generic item", 0, 10));
        app.updateQuality();
        assertEquals(8, app.items[0].quality, "When sellin date is 0 then quality decreases twice as fast");
    }

    @Test
    public void GenericItem_QualityNeverGoesNegative() {
        GildedRose app = new GildedRose(createItemArray("generic item", 0, 0));
        app.updateQuality();
        assertEquals(0, app.items[0].quality, "Quality will not go negative once it is zero");
    }

    @Test
    public void AgedBrie_QualityIncreases() {
        GildedRose app = new GildedRose(createItemArray("Aged Brie", 5, 30));
        app.updateQuality();
        assertEquals(31, app.items[0].quality, "Aged Brie increases quality with age");
    }

    @ParameterizedTest(name = "{0} item Quality is Capped at 50")
    @CsvSource({"Aged Brie", BACKSTAGE_PASS})
    public void NonLegendaryItem_ThatImprovesWithAge_QualityIsCappedAt50(String itemName) {
        GildedRose app = new GildedRose(createItemArray(itemName, 5, 50));
        app.updateQuality();
        assertEquals(50, app.items[0].quality, "Quality has an upper limit that is not exceeded");
    }


    @Test
    public void AgedBrie_QualityIncreases_EvenAfterSellInDate() {
        GildedRose app = new GildedRose(createItemArray("Aged Brie", -1, 20));
        app.updateQuality();
        assertEquals(22, app.items[0].quality, "Aged Brie improves twice as fast after sellin date (BUG?)");
    }

    @Test
    public void AgedBrie_QualityIsCappedAt50_EvenWhenReallyOld() {
        GildedRose app = new GildedRose(createItemArray("Aged Brie", -99, 50));
        app.updateQuality();
        assertEquals(50, app.items[0].quality, "Quality has an upper limit, even when cheese is old");
    }

    @Test
    public void BackstagePass_QualityIncreasesEachDay() {
        GildedRose app = new GildedRose(createItemArray(BACKSTAGE_PASS, 30, 23));
        app.updateQuality();
        assertEquals(24, app.items[0].quality, "Backstage Pass increases quality with age");
    }

    @Test
    public void BackstagePass_QualityIncreasesMoreAsConcertNears() {
        GildedRose app = new GildedRose(createItemArray(BACKSTAGE_PASS, 10, 40));
        app.updateQuality();
        assertEquals(42, app.items[0].quality, "Backstage Pass quality increases more when concert is near");
    }

    @Test
    public void BackstagePass_QualityIncreasesMuchMoreWhenConcertIsClose() {
        GildedRose app = new GildedRose(createItemArray(BACKSTAGE_PASS, 5, 40));
        app.updateQuality();
        assertEquals(43, app.items[0].quality, "Backstage Pass quality increases even more when concert is almost here");
    }

    @Test
    public void BackStagePass_QualityDropsToZeroWhenConcertPasses() {
        GildedRose app = new GildedRose(createItemArray(BACKSTAGE_PASS, 0, 50));
        app.updateQuality();
        assertEquals(0, app.items[0].quality, "Backstage Pass is worthless when concert has passed");
    }

    //TODO: NEW BEHAVIOR
    // conjured items
}

File: csharp/GildedRose/GildedRoseTest.cs
using NUnit.Framework;
using System.Collections.Generic;

namespace csharp
{
    [TestFixture]
    public class GildedRoseTest
    {
        private const string BACKSTAGE_PASS = "Backstage passes to a TAFKAL80ETC concert";

        private List<Item> createItemList(string itemName, int sellIn, int quality) {
            return new List<Item> { new Item { Name = itemName, SellIn = sellIn, Quality = quality } };
        }        

        [Test]
        public void LegendaryItem_QualityDoesNotDecrease() {
            // Arrange
            GildedRose sut = new GildedRose(createItemList("Sulfuras, Hand of Ragnaros", 20, 80));
    
            // Act
            sut.UpdateQuality();
    
            // Assert
            Assert.AreEqual(80, sut.Items[0].Quality, "Quality is not decreased for this legendary item");
        }
    
        [Test]
        public void LegendaryItem_NeverHasToBeSold() {
            GildedRose sut = new GildedRose(createItemList("Sulfuras, Hand of Ragnaros", 1, 80));
            sut.UpdateQuality();
            Assert.AreEqual(1, sut.Items[0].SellIn, "Sellin is not decreased for this legendary item");
        }

        [TestCase("generic item")]
        [TestCase("Aged Brie")]
        [TestCase(BACKSTAGE_PASS)]
        public void NonLegendaryItem_SellInDate_Decreases(string name) {
            GildedRose sut = new GildedRose(createItemList(name, 8, 10));
            sut.UpdateQuality();
            Assert.AreEqual(7, sut.Items[0].SellIn, "Item sellin date should decrease by 1 each day");
        }
    
        [Test]
        public void GenericItem_SellinDateGoesNegative() {
            GildedRose sut = new GildedRose(createItemList("generic item", 0, 25));
            sut.UpdateQuality();
            Assert.AreEqual(-1, sut.Items[0].SellIn, "Sellin date will go negative once sellin date is reached");
        }
    
        [Test]
        public void GenericItem_QualityDecreasesBeforeSellinDate() {
            GildedRose sut = new GildedRose(createItemList("generic item", 5, 10));
            sut.UpdateQuality();
            Assert.AreEqual(9, sut.Items[0].Quality, "Item quality should only decrease by 1 each day");
        }
    
        [Test]
        public void GenericItem_QualityDecreasesTwiceAsFastAfterSellinDate() {
            GildedRose sut = new GildedRose(createItemList("generic item", 0, 10));
            sut.UpdateQuality();
            Assert.AreEqual(8, sut.Items[0].Quality, "When sellin date is 0 then quality decreases twice as fast");
        }
    
        [Test]
        public void GenericItem_QualityNeverGoesNegative() {
            GildedRose sut = new GildedRose(createItemList("generic item", 0, 0));
            sut.UpdateQuality();
            Assert.AreEqual(0, sut.Items[0].Quality, "Quality will not go negative once it is zero");
        }
    
        [Test]
        public void AgedBrie_QualityIncreases() {
            GildedRose sut = new GildedRose(createItemList("Aged Brie", 5, 30));
            sut.UpdateQuality();
            Assert.AreEqual(31, sut.Items[0].Quality, "Aged Brie increases quality with age");
        }
    
        [Test]
        public void AgedBrie_QualityIsCappedAt50() {
            GildedRose sut = new GildedRose(createItemList("Aged Brie", 5, 50));
            sut.UpdateQuality();
            Assert.AreEqual(50, sut.Items[0].Quality, "Quality has an upper limit that is not exceeded");
        }
    
        [Test]
        public void AgedBrie_QualityIncreases_EvenAfterSellInDate() {
            GildedRose sut = new GildedRose(createItemList("Aged Brie", -1, 20));
            sut.UpdateQuality();
            Assert.AreEqual(22, sut.Items[0].Quality, "Aged Brie improves twice as fast after sellin date (BUG?)");
        }
    
        
        [Test]
        public void AgedBrie_QualityIsCappedAt50_EvenWhenReallyOld() {
            GildedRose sut = new GildedRose(createItemList("Aged Brie", -99, 50));
            sut.UpdateQuality();
            Assert.AreEqual(50, sut.Items[0].Quality, "Quality has an upper limit, even when cheese is old");
        }
    
        [Test]
        public void BackstagePass_QualityIncreasesEachDay() {
            GildedRose sut = new GildedRose(createItemList(BACKSTAGE_PASS, 30, 23));
            sut.UpdateQuality();
            Assert.AreEqual(24, sut.Items[0].Quality, "Backstage Pass increases quality with age");
        }
    
        [Test]
        public void BackstagePass_QualityIncreasesMoreAsConcertNears() {
            GildedRose sut = new GildedRose(createItemList(BACKSTAGE_PASS, 10, 40));
            sut.UpdateQuality();
            Assert.AreEqual(42, sut.Items[0].Quality, "Backstage Pass quality increases more when concert is near");
        }
    
        [Test]
        public void BackstagePass_QualityIncreasesMuchMoreWhenConcertIsClose() {
            GildedRose sut = new GildedRose(createItemList(BACKSTAGE_PASS, 5, 40));
            sut.UpdateQuality();
            Assert.AreEqual(43, sut.Items[0].Quality, "Backstage Pass quality increases even more when concert is almost here");
        }
    
        [Test]
        public void BackstagePass_QualityIsCappedAt50() {
            GildedRose sut = new GildedRose(createItemList(BACKSTAGE_PASS, 40, 50));
            sut.UpdateQuality();
            Assert.AreEqual(50, sut.Items[0].Quality, "Quality has an upper limit that is not exceeded");
        }
    
        [Test]
        public void BackStagePass_QualityDropsToZeroWhenConcertPasses() {
            GildedRose sut = new GildedRose(createItemList(BACKSTAGE_PASS, 0, 50));
            sut.UpdateQuality();
            Assert.AreEqual(0, sut.Items[0].Quality, "Backstage Pass is worthless when concert has passed");
        }

        //NEW BEHAVIOR
        // conjured items
    
    }
}
